shared_task_code/ecs_task_queue_consumer.py

```diff
@@ -0,0 +1,131 @@
+                        },
+                        {
+                            "name": "RESULTS_BUCKET",
+                            "value": ssm_params[self.RESULTS]
+                        },
+                        {
+                            "name": "S3_METADATA",
+                            "value": "scan_end_time="+task_params['scan_end_time'] +
+                            ",address="+task_params['address'] +
+                            ",address_type="+task_params['address_type']
+
+
+                        }]
+                }]
+            }
+        }
+        print(f"Submitting task: {dumps(ecs_params)}")
+        task_response = self.ecs_client.run_task(**ecs_params)
+        print(f"Submitted scanning task: {dumps(task_response)}")
+
+        failures = task_response["failures"]
+        if len(failures) != 0:
+            raise RuntimeError(
+                f"ECS task failed to start {dumps(failures)}")
+
+        def run_scan(self, scan, message_id):
+            # Use this format to have uniformly named files in S3
+            date_string = f"{datetime.now():%Y-%m-%dT%H%M%S%Z}"
+            s3file = f"{message_id}-{date_string}-{self.task_name}.txt"
+
+            results_filename = f"/tmp/{s3file}"
+
+            if self.func_taskcode != None:
+                self.func_taskcode(self.event, scan, message_id, results_filename, '/tmp/')
+
+            subprocess.check_output(f'cd /tmp;tar -czvf "{s3file}.tar.gz" "{s3file}"', shell=True)
+            s3 = boto3.resource("s3", region_name=self.region)
+            s3.meta.client.upload_file(
+                f"/tmp/{s3file}.tar.gz",
+                self.event["ssm_params"][self.RESULTS],
+                f"{s3file}.tar.gz",
+                ExtraArgs={'ServerSideEncryption': "AES256", 'Metadata': scan})
+
+    def start(self, ValidateData=None, TaskCode=None):
+        # Pass in variables by name for:
+        # ValidateData=func() - optional (validates the event data)
+        # TaskCode=func() - the code to execute for this task
+        self.event['ssm_params'] = self.get_ssm_params(
+            self.ssm_client, [self.PRIVATE_SUBNETS,
+                              self.SUBNETS, self.CLUSTER, self.RESULTS, self.SECURITY_GROUP, self.IMAGE_ID])
+
+        self.func_validatedata = ValidateData
+        self.func_taskcode = TaskCode
+        for record in self.event["Records"]:
+            # scan = loads(record["body"])
+            scan = record["body"]
+            message_id = f"{record['messageId']}"
+            if self.func_validatedata != None:
+                # validatedata should validate the data, and return a cleaned/destructured
+                # version of it, e.g. for nmap scanner there's a case where there are a list
+                # of things to scan - so in this case, the string 'scan', could return as
+                # an array of strings
+                (valid, scan) = self.func_validatedata(self.event, scan, message_id)

```

Comment:
Can we use a different name for the destructured scan, as opposed to the input scan?

Reply:
this intentionally overwrites scan, as the rest of the function uses it, and only goes into this line if a validation function was provided